src/ripple/app/tx/impl/Transactor.cpp

```diff
@@ -456,6 +593,11 @@ Transactor::apply()
 NotTEC
 Transactor::checkSign(PreclaimContext const& ctx)
 {
+    // hook emitted transactions do not have signatures
+    if (ctx.view.rules().enabled(featureHooks) &&
```

Comment:
This is a really scary bit of code here. We need to be 1000% sure that a transaction that passes this `if` can **NEVER** be generated from something other than a hook under **ANY** circumstances. 

It might actually be nice to have a test _here_ that verifies that the account has authorized the hook that issued the transaction to issue transactions on its behalf.

Reply:
Definitely agree that this is a dangerous and pivotal piece of code.

The `isEmittedTxn` function tests for the presence of the `sfEmitDetails` block in a transaction. Provided there's no way to put that validly into a non-emitted transaction then this condition will never pass. This is already blocked at the netcode in 5 different locations... however given how cheap it is to check, the more checks the better. (Ultimately it could actually be checked in TxQ. Unless it's added from the emitDir it should never have sfEmitDetails.)

It is possible to check if the Hook that emitted the txn was authorized by the emitting account, but there's an annoying corner case here: since the emitted txn is "from the past" i.e. created by a past execution of a hook, the current state of the ledger might not be the same as it was then. That is: the emitted txn was authorized when it was emitted but at the time the signature is being checked (at insertion into the next ledger, or the one after that) it might not be. And a very real case of this is when a Hook emits a txn that deletes itself. The "one and done" self deleting hook is actually a very useful construct, and not something we want to accidentally prevent.

We could just check the historic ledger. But if we do this then we need to ensure a validator resyncing has at least that much history before this code ever executes. (Even though it's only 5 ledgers, I'm not sure if that is necessarily the case at the moment?)